Add vector search tuning tips (#1196)
hazelcast/hz-docs: docs/modules/data-structures/pages/vector-search-overview.adoc

To decrease pressure on heap memory, you can decrease the number of parallel migrations using `hazelcast.partition.max.parallel.migrations` and `hazelcast.partition.max.parallel.replications`.
====

== Tuning tips

> 1. In Hazelcast 5.5 for searches with small `topK` (e.g. 10) it may be beneficial to artificially increase `topK`, adjust `partitionLimit` accordingly and discard extra results. For initial `topK=10` a good starting point could be `topK` increased to 100 and `partitionLimit` between 50 and 100. While this will make the search slower, it will also improve quality, sometimes significantly. Overall this setup can be more efficient than increasing index build parameters (`max-degree`, `ef-construction`) which results in slower index build and searches. With very small `topK` or `paritionLimit` the search algorithm is less able to escape local minima and find best results.

Reviewer: ```suggestion
1. In Hazelcast 5.5 for searches with small `topK` (for example, 10) it may be beneficial to artificially increase `topK`, adjust `partitionLimit` accordingly, and discard extra results. While `topK=10` is a good starting point, you could increase `topK` to 100 and set `partitionLimit` to between 50 and 100. While this will make the search slower, it will also improve quality, sometimes significantly. Overall, this setup can be more efficient than increasing index build parameters (`max-degree`, `ef-construction`) which results in slower index build and searches. With a very small `topK` or `paritionLimit`, the search algorithm is less able to escape local minima and find the best results.
```

Author: ```suggestion
1. In Hazelcast 5.5 for searches with small `topK` (for example, 10) it may be beneficial to artificially increase `topK`, adjust `partitionLimit` accordingly, and discard extra results. If you need 10 results a good starting point for tuning could be `topK=100` and `partitionLimit` between 50 and 100. While this will make the search slower, it will also improve quality, sometimes significantly. Overall, this setup can be more efficient than increasing index build parameters (`max-degree`, `ef-construction`) which results in slower index build and searches. With a very small `topK` or `paritionLimit`, the search algorithm is less able to escape local minima and find the best results.
```

my updated suggestion. The idea is that if you need 10 results, request 100 and discard 90 and use that setup as starting point for further tuning. This is counterintuitive and we should improve that in the next version

Reviewer: Fantastic @k-jamroz, yes, go with your updated suggestion. Thank you.

> 2. Vector deduplication does not incurr significant overhead for upload (usually less than 1%) and search. You may consider disabling it to get slighly better performance and smaller memory usage if your dataset does not contain duplicated vectors. However, beware that in presence of many duplicated vectors and deduplication disabled similarity search may return poor quality results.

Reviewer: ```suggestion
2. Vector deduplication does not incur significant overhead for uploads (usually less than 1%) and searches. You may consider disabling it to get slightly better performance and smaller memory usage if your dataset does not contain duplicated vectors. However, beware that in the presence of many duplicated vectors with deduplication disabled, a  similarity search may return poor quality results.
```

> 1. In Hazelcast 5.5 for searches with small `topK` (for example, 10) it may be beneficial to artificially increase `topK`, adjust `partitionLimit` accordingly, and discard extra results. If you need 10 results a good starting point for tuning could be `topK=100` and `partitionLimit` between 50 and 100. While this will make the search slower, it will also improve quality, sometimes significantly. Overall, this setup can be more efficient than increasing index build parameters (`max-degree`, `ef-construction`) which results in slower index build and searches. With a very small `topK` or `paritionLimit`, the search algorithm is less able to escape local minima and find the best results.
> 2. Vector deduplication does not incur significant overhead for uploads (usually less than 1%) and searches. You may consider disabling it to get slightly better performance and smaller memory usage if your dataset does not contain duplicated vectors. However, beware that in the presence of many duplicated vectors with deduplication disabled, a  similarity search may return poor quality results.

Reviewer 2: ```suggestion
1. For searches with small `topK` (for example, 10) it may be beneficial to artificially increase `topK`, adjust `partitionLimit` accordingly, and discard extra results. If you need 10 results, a good starting point for tuning could be `topK=100` and a `partitionLimit` between 50 and 100. While this will make the search slower, it will also improve quality, sometimes significantly. Overall, this setup can be more efficient than increasing index build parameters (`max-degree`, `ef-construction`) which results in slower index builds and searches. With a very small `topK` or `paritionLimit`, the search algorithm is less able to escape local minima and find the best results.
2. Vector deduplication does not incur significant overhead for uploads (usually less than 1%) and searches. You may consider disabling it to get slightly better performance and smaller memory usage if your dataset does not contain duplicated vectors. However, be aware that in the presence of many duplicated vectors with deduplication disabled, a  similarity search may return poor quality results.
```
